src/easylink/configuration.py

```diff
@@ -14,29 +15,68 @@
 ENVIRONMENT_ERRORS_KEY = "ENVIRONMENT ERRORS"
 
 DEFAULT_ENVIRONMENT = {
-    "computing_environment": "local",
-    "container_engine": "undefined",
-    "implementation_resources": {
-        "memory": 1,  # GB
-        "cpus": 1,
-        "time_limit": 1,  # hours
-    },
-    "spark": {
-        "workers": {
-            "num_workers": 2,
-            "cpus_per_node": 1,
-            "mem_per_node": 1,  # GB
+    "environment": {
+        "computing_environment": "local",
+        "container_engine": "undefined",
+        "implementation_resources": {
+            "memory": 1,  # GB
+            "cpus": 1,
             "time_limit": 1,  # hours
         },
-        "keep_alive": False,
+    }
+}
+SPARK_DEFAULTS = {
+    "workers": {
+        "num_workers": 2,
+        "cpus_per_node": 1,
+        "mem_per_node": 1,  # GB
+        "time_limit": 1,  # hours
     },
+    "keep_alive": False,
 }
 
 # Allow some buffer so that slurm doesn't kill spark workers
 SLURM_SPARK_MEM_BUFFER = 500
 
 
-class Config:
+def load_params_from_specification(
+    pipeline_specification: str, input_data: str, computing_environment: str, results_dir: str
+) -> Dict:
+    return {
+        "pipeline": load_yaml(pipeline_specification),
+        "input_data": load_input_data_paths(input_data),
+        "environment": load_computing_environment(computing_environment),
+        "results_dir": Path(results_dir),
+    }
+
+
+def load_input_data_paths(input_data_specification_path: Union[str, Path]) -> List[Path]:
+    input_data_paths = load_yaml(input_data_specification_path)
+    if not isinstance(input_data_paths, dict):
+        raise TypeError(
+            "Input data should be submitted like 'key': path/to/file. "
+            f"Input was: '{input_data_paths}'"
+        )
+    file_list = [Path(filepath).resolve() for filepath in input_data_paths.values()]
+    return file_list
+
+
+def load_computing_environment(
+    computing_environment_specification_path: Optional[str],
+) -> Dict[Any, Any]:
+    """Load the computing environment yaml file and return the contents as a dict."""
+    if not computing_environment_specification_path:
+        return {}  # handles empty environment.yaml
+    elif not Path(computing_environment_specification_path).is_file():
+        raise FileNotFoundError(
+            "Computing environment is expected to be a path to an existing"
+            f" yaml file. Input was: '{computing_environment_specification_path}'"
+        )
+    else:
+        return load_yaml(computing_environment_specification_path)
+
+
+class Config(LayeredConfigTree):
```

Comment:
Is it better to inherit from `LayeredConfigTree` or just to have an attribute that is one that is referenced? I'm not sure how we're interacting with this object.

Reply:
I considered both ways, but to me it seems like the configuration should be a static, frozen configtree after we construct and validate it in `__init__()`. All we should be doing is accessing its nested keys, and I didn't see a strong reason why having (for example) a configtree as attribute for each of (pipeline, environment, input_data) would be better than just having one ConfigTree with top-level keys for each.